Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+            // It is at this point we want to say the collection has been uploaded.
+            return self.uploadBatches(token, batches: batch) >>> {
+                return self.finishBatch(token, lines: lastBatch)
+                    >>== effect({ onCollectionUploaded($0.value) })
+                    >>> succeed
+            }
+        }
+    }
+
+    private func startBatch(lines: [String]) -> Deferred<Maybe<BatchToken>> {
+        let batchStartParam = NSURLQueryItem(name: "batch", value: "true")
+
+        // Attempt to upload some records and see if we get back a token we can use for batches.
+        return self.uploader(lines: lines, ifUnmodifiedSince: ifUnmodifiedSince, queryParams: [batchStartParam]) >>== { storageResponse in
+            if let batchToken = storageResponse.value.batchToken {
+                log.debug("Uploaded \(lines.count) records and received batch token \(batchToken)")
+                return deferMaybe(batchToken)
+            } else {
+                log.debug("Uploaded \(lines.count) records but received no batch token")
+                return deferMaybe(BatchingNotSupported(response: storageResponse.value))
+            }
+        }
+    }
+
+    private func uploadBatches(token: BatchToken, batches: [[String]]) -> Success {
+        let batchParam = NSURLQueryItem(name: "batch", value: token)
+        let uploadBatch: (lines: [String]) -> Success = { lines in
+            return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: [batchParam])
+                >>> effect({ log.debug(("Uploaded \(lines.count) records for batch \(token)")) })
+        }
+
+        return walk(batches, f: uploadBatch)
+    }
+
+    private func finishBatch(token: BatchToken, lines: [String]) -> Deferred<Maybe<StorageResponse<POSTResult>>> {
+        let batchParam = NSURLQueryItem(name: "batch", value: token)
+        let commitParam = NSURLQueryItem(name: "commit", value: "true")
+        return self.uploader(lines: lines, ifUnmodifiedSince: ifUnmodifiedSince, queryParams: [batchParam, commitParam])
+    }
+
+    func batchesFromRecords(records: [Record<T>]) -> Maybe<[[String]]> {
+
+        // Place the record ID alongside the line text for error reporting later on
+        let idWithLine: (record: Record<T>) -> (GUID, String)? = { record in
+            guard let line = self.serializeRecord(record) else {
+                return nil
+            }
+            return (record.id, line)
+        }
+
+        let pairs : [(GUID, String)] = records.flatMap(idWithLine)
+
+        var batches = [[String]]()
+        var batch = [String]()
+        var size = 0
+
+        for (id, line) in pairs {
+            let lineSize = line.utf8.count
+
+            if lineSize > Sync15StorageClient.maxRecordSizeBytes {
+                return Maybe(failure: RecordTooLargeError(size: lineSize, guid: id))
```

Comment:
Do this check waaaaaay earlier: at the very moment you try to add the record to the batcher! I shouldn't get a delayed failure in `.commit` when I can get a synchronous failure at the moment of addition.